pulp/apis/cuopt_api.py

```diff
@@ -0,0 +1,337 @@
+import ctypes
+import os
+import subprocess
+import sys
+import warnings
+from uuid import uuid4
+import numpy as np
+from ..constants import (
+    LpBinary,
+    LpConstraintEQ,
+    LpConstraintGE,
+    LpConstraintLE,
+    LpContinuous,
+    LpInteger,
+    LpMaximize,
+    LpMinimize,
+    LpStatusInfeasible,
+    LpStatusNotSolved,
+    LpStatusOptimal,
+    LpStatusUnbounded,
+    LpStatusUndefined,
+)
+from .core import (
+    LpSolver,
+    LpSolver_CMD,
+    PulpSolverError,
+    clock,
+    ctypesArrayFill,
+    sparse,
+)
+
+# COPT string convention
+if sys.version_info >= (3, 0):
+    coptstr = lambda x: bytes(x, "utf-8")
+else:
+    coptstr = lambda x: x
+
+byref = ctypes.byref
+
+# Constraint Sense Converter
+sense_conv = {LpConstraintLE: "L",
+              LpConstraintGE: "G",
+              LpConstraintEQ: "E",
+             }
+
+def process_constraint(args):
+    local_data = []
+    local_vars = []
+    local_sense = []
+    local_rhs = []
+    for constraints in args:
+        name, constraint = constraints
+        local_data.extend(list(constraint.values()))
+        local_vars.extend(list(constraint.keys()))
+        local_sense.append(sense_conv[constraint.sense])
+        local_rhs.append(-constraint.constant)
+    return local_data, local_vars, local_sense, local_rhs
+
+
+def process_variables(var):
+    #obj_coeff = lp.objective.get(var, 0.0))
+    lowBound = var.lowBound
+    if lowBound is None:
+        lowBound = -np.inf
+    upBound = var.upBound
+    if upBound is None:
+        upBound = np.inf
+    varType = "C"
+    if var.cat == LpInteger and self.mip:
+        varType = "I"
+    if var.cat == LpBinary and self.mip:
+        varType = "I"
+        lowBound = 0
+        upBound  = 1
+    return lowBound, upBound, varType, var.name
+
+
+class CUOPT(LpSolver):
+    """
+    The CUOPT Optimizer via its python interface
+    """
+
+    name = "CUOPT"
+
+    try:
+        global cuopt
+        import cuopt  # type: ignore[import-not-found]
+    except:
+
+        def available(self):
+            """True if the solver is available"""
+            return False
+
+        def actualSolve(self, lp, callback=None):
+            """Solve a well formulated lp problem"""
+            raise PulpSolverError("CUOPT: Not available")
+
+    else:
+
+        def __init__(
+            self,
+            mip=True,
+            msg=True,
+            timeLimit=None,
+            gapRel=None,
+            warmStart=False,
+            logPath=None,
+            **solverParams,
```

Comment:
solverParams is not defined in the docstring

Reply:
Added docstring in #839